README
Nginx::Simple version 0.06
===========================

See the individual module documentation for more information

INSTALLATION

To install this module type the following:

    perl Makefile.PL
    make
    make test
    make install

COPYRIGHT AND LICENCE

Copyright (C) 2010 Michael J. Flickinger <mjflick@open-site.org>

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


Changes
Revision history for Perl extension for Nginx::Simple

0.06 2010-12-22
    - Added lightweight dispatching engine along with "ngs" utility.

0.04 2008-12-08
    - fixing status bug

0.03 2008-12-08
    - Added cleanup sub.
    - Ridded need of handler sub in primary module.

0.02 2008-12-05
    - New release.

0.01 2008-09-28
    - Initial import.


lib/Nginx/Simple/Ajax.pm
package Nginx::Simple::Ajax;

use strict;

use XML::Simple;

=head1 MODULE

Nginx::Simple::Ajax

=head1 METHODS

=cut

=head3 my $ajax = new Ajax($self);

Initialize with Nginx::Simple object.

=cut

sub new 
{
    my ($class, $nginx_simple) = @_;
    my $self = {};
    
    $self->{export}{javascript} = q[];
    $self->{nginx_simple} = $nginx_simple;

    bless($self);

    return $self;
}

=head3 $self->read

Read values into an ajax object.

=cut

sub read
{
    my $self = shift;

    my %items_table;
    my $form_data = $self->{nginx_simple}->request_body;

    return unless $form_data;

    $self->reset;
    
    my $xml = eval { 
        XMLin($form_data, ForceArray => [ 'element' ])
    };

    return unless $xml;

    ### read in and parse inputs from server
    for my $element (ref($xml->{element}) eq 'ARRAY' 
                     ? (@{$xml->{element}})
                     : ($xml->{element}))
    {
        $items_table{$_} = $element->{$_}{content}
            for keys %$element;
    }

    $self->{items_table}  = \%items_table;
}

=head3 $self->param( $id )

Akin to Nginx::Simple->param.

=cut

sub param
{
    my ($self, $id) = @_;

    return unless ref $self->{items_table} eq 'HASH';

    if ($id)
    {
        return $self->get( id => $id );
    }
    else
    {
        return keys %{$self->{items_table}};
    }
}   

=head3 $self->get( id => 'foo' )

Returns data as scalar given an element_id.

=cut

sub get
{
    my ($self, %params) = @_;

    die qq(\$self->get: Param id must be given\n)
        unless $params{id};

    my $item_id = $params{id};
    my $data = $self->{items_table}{$item_id};

    return $data;

}

=head3 $self->add( foo => 'bar' || javascript => "alert('foo!')" )

Enqueue an item to be generated into javascript xml.

=cut

sub add
{
    my ($self, %params) = @_;

    my $export = $self->{export};

    if ( exists( $params{javascript} ) )
    {
        $export->{javascript} .= $params{javascript};
    }

    if ( exists( $params{pre_javascript} ) )
    {
        $export->{pre_javascript} .= $params{pre_javascript};
    }

    for my $key (keys %params)
    {
        next if $key eq 'javascript';
        next if $key eq 'pre_javascript';
        
        $export->{elements}{$key} = $params{$key};
    }   
}

=head3 $self->send

Send Ajax returned XML.

=cut

sub send
{
    my $self = shift;

    my $export = $self->{export};

    $self->{nginx_simple}->header_set('Content-Type', 'text/xml');

    my $xml = qq(<x-response>\n);
    
    if ($export->{javascript})
    {
        my $js = $export->{javascript};

        $xml .= qq{ <response type="javascript"><![CDATA[$js]]></response>\n};
    }

    if ($export->{pre_javascript})
    {
        my $js = $export->{pre_javascript};

        $xml .= qq{ <response type="pre-javascript"><![CDATA[$js]]></response>\n};
    }

    for my $key (keys %{$export->{elements}})
    {
        my $html = $export->{elements}{$key};
        $xml .= qq{ <response type="element" id="$key"><![CDATA[$html]]></response>\n};
    }
    $xml .= qq(</x-response>\n);

    $self->{nginx_simple}->print($xml);
}

=head3 $self->reset

Reset the object.

=cut

sub reset
{
    my $self = shift;

    delete $self->{items_table};
    delete $self->{export};

    $self->{export}{javascript}     = q[];
    $self->{export}{pre_javascript} = q[];
}

=head1 Author

Michael J. Flickinger, C<< <mjflick@open-site.org> >>

=head1 Copyright & License

You may distribute under the terms of either the GNU General Public
License or the Artistic License.

=cut

1;


lib/Nginx/Simple/Cookie.pm
package Nginx::Simple::Cookie;

use strict;

use CGI::Cookie;

=head1 MODULE

Nginx::Simple::Cookie

=head1 METHODS

=over 4

=cut

=item $self->new

Return cookie dispatcher.

=cut

sub new
{
    my ($class, $ns) = @_;
    my $self = { nginx_simple => $ns };
    bless($self);
    return $self;
}

=item Cookie->set

Set cookie.

=cut

sub set 
{
    my ($self, %params) = @_;

    $self->{nginx_simple}->header_set('Set-Cookie', new CGI::Cookie(%params));
}

=item my %params = Cookie->read

Read all cookies.

=cut

sub read
{
    my ($self, %params) = @_;

    my $cookies = $self->{nginx_simple}->header_in('Cookie');
    
    return parse CGI::Cookie($cookies);
}

=head1 Author

Michael J. Flickinger, C<< <mjflick@open-site.org> >>

=head1 Copyright & License

You may distribute under the terms of either the GNU General Public
License or the Artistic License.

=cut

1;


lib/Nginx/Simple/Dispatch.pm
package Nginx::Simple::Dispatch;

use Exporter;
use strict;

use Nginx::Simple;
use Nginx::Simple::Dispatcher;

# Configure exporter.
our @ISA    = qw(Exporter Nginx::Simple::Dispatcher);
our @EXPORT = (@Nginx::Simple::HTTP_STATUS_CODES);

our $VERSION = '0.01';

=head1 NAME

Nginx::Simple::Dispatch - Easy to use dispatcher interface for "--with-http_perl_module"

=head1 SYNOPSIS

MainPage.pm:

   package MainPage;

   # app_path is where the application's URI begins
   # auto_import allows for lazyloading of application modules
   use Nginx::Simple::Dispatch( app_path => '/foo', auto_import => 1 );
   use strict;

   sub index :Index {
       my $self = shift;

       $self->print('I am the main page people will see.');
   }

   sub about :Action {
       my $self = shift;

       $self->print('Let us pretend the about page goes here, shall we?');
   }

   sub error {
       my $self = shift;
       my $error = shift;

       $self->status(500);
       $self->print('Something blew up! Kaboooooom!<hr>');
       $self->print("Details: <pre>$error</pre>");
   }

   sub bad_dispatch {
       my $self = shift;
       my $error = shift;

       $self->status(404);
       $self->print('Page not found.');
   }

   # do something after we do something
   sub cleanup
   {
       warn "[looging dragons: good]\n";
   }

   1;

MainPage/weather.pm

   package MainPage::weather;

   use base 'MainPage';
   use strict;

   sub index :Index {
       my $self = shift;

       $self->print("zap zap ook");
   }

   sub corn :Action {
       my $self = shift;

       $self->print("men");
   }

   1;

=cut

sub import
{
    my ($class, %params) = @_;
    my $caller = caller;

    # inject a handler method
    {
        no strict 'refs';

        my $caller_isa = "$caller\::ISA";

        @{$caller_isa} = qw(
            Nginx::Simple
            Nginx::Simple::Dispatcher
        );

        *{"$caller\::handler"} = sub {
            return local_dispatch(
                shift, 
                class         => $caller,
                app_path      => $params{app_path},
                auto_import   => $params{auto_import},
                auto_redirect => $params{auto_redirect},
            );
        };
    }

    __PACKAGE__->export_to_level(1, $class);
}

# where do we dispatch to

sub local_dispatch
{
    my ($self, %params) = @_;
    my $class    = $params{class};
    my $app_path = $params{app_path};
    my $path     = $self->uri;

    # trip the app_path off $path, when applicable
    $path =~ s/^$app_path// if $app_path;

    my $dispatch_data = 
        __PACKAGE__->dig_for_dispatch(
            class       => $class,
            path        => $path,
            auto_import => $params{auto_import},
        );

    # ensure indexes always end in a slash
    if ($params{auto_redirect} and $dispatch_data->{index} and $path !~ /\/$/)
    {
        $self->status(302);
        $self->header_out(Location => $self->uri . '/');
        $self->send_http_header;

        return;
    }

    if ($dispatch_data->{error} eq 'bad_dispatch')
    {
        return dispatch(
            $self, 
            class  => $class,
            method => 'bad_dispatch',
            bless  => 1,
        );
    }
    elsif ($dispatch_data->{error})
    {
        return dispatch(
            $self, 
            class  => $class,
            method => 'error',
            error  => $dispatch_data->{error},
            bless  => 1,
        );
    }
    else # prepare to dispatch for real
    {
        return dispatch(
            $self,
            class      => $dispatch_data->{class},
            method     => $dispatch_data->{method},
            base_class => $class,
            bless      => 1,
        );
    }
}

=head1 Author

Michael J. Flickinger, C<< <mjflick@open-site.org> >>

=head1 Copyright & License

You may distribute under the terms of either the GNU General Public
License or the Artistic License.

=cut

1;


lib/Nginx/Simple/Dispatcher.pm
package Nginx::Simple::Dispatcher;

use base qw(
    Exporter
    Nginx::Simple::Dispatcher::Attributes
);

use strict;

our @EXPORT = qw(dig_for_dispatch);

# keep a list of dispatched paths
my %path_cache;

=head1 NAME 

Nginx::Simple::Dispatcher

=head1 Synopsis

Automatic dispatcher built on code attributes.

=head1 Methods

=cut

sub dig_for_dispatch
{
    my ($self, %params) = @_;
    my $class       = $params{class};
    my $path        = $params{path};
    my $called_path = $path || $params{called_path}; # the complete path

    # reset self
    $self = $params{self} if $params{self};

    # dereference class
    $class    = ref $class ? ref $class : $class;

    # sanitize path
    {
        # remove starting slash
        $path =~ s/^\///;

        # remove trailing slash
        $path =~ s/\/+$//;

        # remove index trailing (you can't call index directly)
        $path =~ s/(\/|^)index$//;

        # append 'index' if no path given
        $path .= 'index' unless $path;
    }

    # check cache (for fast dispatches)
    return $path_cache{$called_path}
        if $path_cache{$called_path};

    # build list of paths
    my @paths = split('/', $path);

    # build method call
    my $call_class   = $class;
    my $method       = pop @paths;
    my $remote_class = $class;
       $remote_class = join('::', $class, join('::', @paths))
           if @paths;

    &_class_is_imported($remote_class) if $params{auto_import};

    if (UNIVERSAL::can($remote_class, 'get_dispatch_flags'))
    {
        my $methods    = $remote_class->get_dispatch_flags;
        my $used_index = 0;

        # if this is a page index, find the index sub name
        if ($method eq 'index')
        {
            ($method) = grep { $methods->{$_} eq 'index' } keys %$methods;
            $used_index = 1;
        }

        if ($methods->{$method})
        {
            # perform dispatch
            {
                no strict 'refs'; # evil
                my $subptr = join('::', $remote_class, $method);

                # store path in cache
                $path_cache{$called_path} = {
                    class  => $remote_class,
                    method => $method,
                    sub    => \&$subptr,
                    index  => $used_index,
                };

                return $path_cache{$called_path};
            }
        }
        else
        {
            my $jump_class = join('::', $remote_class, $method);

            &_class_is_imported($jump_class) if $params{auto_import};

            if (UNIVERSAL::can($jump_class, 'get_dispatch_flags'))
            {
                return $jump_class->dig_for_dispatch(
                    self        => $self,
                    class       => $jump_class,
                    path        => '',
                    called_path => $called_path,
                );
            }

            return { error => 'bad_dispatch' };
        }
    }
    else # get_dispatch_flags is not assessable
    {
        return { error => 'bad_dispatch' };
    }
}

# _class_is_imported(Some::Class)
#
# If a class is not imported, import it.
#

sub _class_is_imported
{
    my $class = shift;

    (my $class_file = $class) =~ s{::}{/}g;
    $class_file .= '.pm'; # Let's assume all class files end in .pm

    if (not exists $INC{$class_file})
    {
        for my $base_path (@INC)
        {
            my $full_path = join('/', $base_path, $class_file);
            
            if (-e $full_path)
            {
                eval qq{use $class;};
                warn "-->$@<--" if $@;
                return;
            }
        }        
    }
}

=head1 Author

Michael J. Flickinger, C<< <mjflick@open-site.org> >>

=head1 Copyright & License

You may distribute under the terms of either the GNU General Public
License or the Artistic License.

=cut

1;



lib/Nginx/Simple/Dispatcher/Attributes.pm
use attributes;
use warnings;
use strict;

package Nginx::Simple::Dispatcher::Attributes;

=head1 NAME 

Nginx::Simple::Dispatcher::Attributes

Attributes:
Index, Page

=head1 Synopsis

Code attributes.

=head1 Methods

=cut

our ( %dispatch_flags, %cached_flags );

sub MODIFY_CODE_ATTRIBUTES
{
    my ($pack, $ref, @attr) = @_;

    for my $attr ( @attr ) {

        # ensure a $dispatch_flags{$pack} hashref, always
        $dispatch_flags{$pack} = { }
            unless exists $dispatch_flags{$pack};

        if  ($attr eq 'Index')
        {
            $dispatch_flags{$pack}{$ref} = 'index';
        }
        if  ($attr eq 'Action')
        {
            $dispatch_flags{$pack}{$ref} = 'action';
        }
        elsif  ($attr =~ /^Path/)
        {
            my $desc;
            $desc = $1 if $attr =~ /\((.*?)\)$/;

            $dispatch_flags{$pack}{$ref} = "$desc";
       }
    }

    return ();
}

sub FETCH_CODE_ATTRIBUTES { $dispatch_flags{shift}{shift} }

=head2 get_dispatch_flags

Returns all autoflags specific to a package with inheritance.

=cut

sub get_dispatch_flags
{
    my $self  = shift;
    my $class = ref $self ? ref $self : $self;

    $cached_flags{$class} = $class->get_package_dispatch_flags
        unless exists $cached_flags{$class};

    return $cached_flags{$class};
}

=head2 get_package_dispatch_flags

Get autoflags, only specific to the called package.

=cut

sub get_package_dispatch_flags
{
    my $self      = shift;
    my $class     = ref $self ? ref $self : $self;
    my @code_refs = keys %{$dispatch_flags{$class}};

    my %flag_methods;
    {
        no strict 'refs';
        for my $key (keys %{"${class}::"})
        {
            my $code_ref = "${class}::${key}";

            if (defined &$code_ref)
            {
                my $code = \&$code_ref;

                $flag_methods{"$key"} = $dispatch_flags{"$class"}{"$code"}
                    if grep { "$code" eq "$_" } @code_refs;
            }
        }
    }

    return \%flag_methods;
}

=head1 Author

Michael J. Flickinger, C<< <mjflick@open-site.org> >>

=head1 Copyright & License

You may distribute under the terms of either the GNU General Public
License or the Artistic License.

=cut

1;



lib/Nginx/Simple.pm
package Nginx::Simple;
use Exporter;

our $VERSION = '0.06';

@ISA = qw(Exporter);

use nginx;
use strict;

use Nginx::Simple::Cookie;
use Time::HiRes qw(gettimeofday tv_interval);

BEGIN { require 5.008004; }

# Export all @HTTP_STATUS_CODES
our @EXPORT = ('dispatch');

# data for request
my %request_data;
my @error_stack;
my $die_error;

=head1 NAME

Nginx::Simple - Easy to use interface for "--with-http_perl_module"

=head1 SYNOPSIS

  nginx.conf:  
    perl_modules perl;
    perl_require Test.pm;

    server {
       listen       80;
       server_name  localhost;

       location / {
          perl Test::handler; # always enter package::handler
       }
    }

  Test.pm:
    package Test;
    use Nginx::Simple;

    # (optional) triggered before main is run
    sub init
    {
        my $self = shift;

        $self->{user} = 'stinky_pete';
    }

    # automatically dispatches here
    sub main
    {
        my $self = shift;
        
        $self->header_set('Content-Type' => 'text/html');
        $self->print('rock on!');
        $self->print("$self->{user} is using the system");
	
        $self->log('I found a rabbit...');
        
        my $something = $self->param("here");
        $self->print("I found $something...");
    }

    # (optional) triggered after main is run
    sub cleanup
    {
        my $self = shift;

        # do something?
    }

    # (optional) triggered on a server error (otherwise returns a normal 500 error)
    sub error
    {
        my $self = shift;
        my $error = shift;

        $self->status(500);
        $self->print("oh, uh, there is an error! ($error)");
    }
    
=cut

sub import
{
    my ($class, $settings) = @_;

    my $caller = caller;
    {
        no strict 'refs';
        *{"$caller\::handler"} = sub { 
            return dispatch(
                shift, 
                class  => $caller,
                method => 'main',
            );                     
        };
    }
    __PACKAGE__->export_to_level(1, $class);
}

sub dispatch
{
    my ($r, %params) = @_;

    my $class      = $params{class};
    my $method     = $params{method};
    my $error      = $params{error};
    my $bless      = $params{bless};
    my $base_class = $params{base_class};

    $r->variable('call_class',  $class               );
    $r->variable('call_method', $method     || 'main');
    $r->variable('error',       $error      || ''    );
    $r->variable('bless',       $bless      || ''    );
    $r->variable('base_class',  $base_class || ''    );

    unless ($method eq 'error')
    {
        @error_stack = ( );
        $die_error   = q[];
    }

    if (not $error and $r->has_request_body(\&handle_request_body))
    {
        return OK; 
    } 
    else
    {
        return &init_dispatcher($r);
    }

    return OK;
}

=head1 METHODS

=head3 $self->server

Returns the nginx server object.

=cut

sub server           { $request_data{server_object}        }

=head3 $self->uri

Returns the uri.

=cut

sub uri              { shift->server->uri                  }

=head3 $self->filename

Returns the path filename.

=cut

sub filename         { shift->server->filename             }

=head3 $self->request_method

Returns the request_method.

=cut

sub request_method   { shift->server->request_method       }

=head3 $self->remote_addr

Returns the remote_addr.

=cut

sub remote_addr      { shift->server->remote_addr          }

=head3 $self->header_in

Return value of header_in.

=cut

sub header_in        { shift->server->header_in(@_)        }

sub rflush           { shift->server->rflush               }
sub flush            { shift->rflush                       }


=head3 $self->print(...)

Output via http.

=cut

sub print 
{
    my $self = shift;

    $request_data{output} .= join '', @_;
}

=head3 $self->auto_header

Returns true if set, otherwise args 1 sets true and 0 false.

=cut

sub auto_header
{
    my ($self, $arg) = @_;

    if (defined $arg)
    {
        if ($arg)
        {
            delete $request_data{disable_auto_header};
        }
        else	
        {
            $request_data{disable_auto_header} = 1;
        }
    }

    return(not exists $request_data{disable_auto_header});
}

=head3 $self->dispatching

Returns true if we're dispatching actively.

=cut

sub dispatching
{
    my ($self, $arg) = @_;

    if (defined $arg)
    {
        if ($arg)
        {
            delete $request_data{disable_dispatching};
        }
        else	
        {
            $request_data{disable_dispatching} = 1;
        }
    }

    return(not exists $request_data{disable_dispatching});
}

=head3 $self->header_set('header_type', 'value')

Set output header.

=cut

sub header_set 
{
    my ($self, $key, $value) = @_;

    $request_data{headers}{$key} = $value;
}

=head3 $self->header('content-type')

Set content type.

=cut

sub header
{
    my ($self, $value) = @_;

    $self->header_set('Content-Type', $value);
}

=head3 $self->headers

Returns hashref of headers.

=cut

sub headers
{
    my ($self, $value) = @_;

    return $request_data{headers};
}

=head3 $self->location('url')

Redirect to a url.

=cut

sub location         { shift->header_set('Location', shift) }

=head3 $self->status(...)

Set output status... (200, 404, etc...)
If no argument given, returns status.

=cut

sub status 
{
    my ($self, $status) = @_;

    if ($status)
    {
        $request_data{status} = $status;
    }
    else
    {
        return $request_data{status};
    }
}

# map $self->log to print STDERR
sub log              { shift; print STDERR @_;             }

=head3 $self->param(...)

Return a parameter passed via CGI--works like CGI::param.

=cut

sub param 
{
    my ($self, $lookup_key) = @_;
    
    my @values;
    my %seen_hash;
    my $request = $request_data{args};

    if ($request_data{request_parts})
    {
        for my $part (@{$request_data{request_parts}})
        {
            if ($lookup_key)
            {
                push @values, $self->unescape($part->{data})
                    if $lookup_key eq $part->{name};
            }
            else
            {
                next if $seen_hash{$part->{name}}++;
                push @values, $part->{name};
            }
        }
    }
    else
    {
        my @args = split('&', $request);
        for my $arg (@args)
        {
            my ($key, $value) = split('=', $arg);
            
            if ($lookup_key)
            {
                push @values, $self->unescape($value)
                    if $lookup_key eq $key;
            }
            else
            {
                next if $seen_hash{$key}++;
                push @values, $key;
            }
        }
    }
    
    return unless @values;
    
    return (scalar @values == 1 ? $values[0] : @values);
}

=head3 $self->param_hash
    
Return a friendly hashref of CGI parameters.

=cut

sub param_hash
{
    my $self = shift;

    my %param_hash;
    
    for my $key ($self->param)
    {
        next if $param_hash{$key};
        
        my @params = $self->param($key);
        
        if (scalar @params == 1)
        {
            $param_hash{$key} = $params[0];
        }
        else
        {
            $param_hash{$key} = [ @params ],
        }
    }
    
    return \%param_hash;
}

=head3 $self->request_body & $self->request
    
Returns request body.

=cut

sub request_body { $request_data{request} }
sub request      { shift->request_body    }

=head3 $self->args

Returns args.

=cut

sub args         { $request_data{args}    }

sub handle_request_body
{
    my $r = shift;

    my $request_body = $r->request_body;

    my @r_body = split("\n", $request_body);

    my %params;

    if (scalar(@r_body) == 1)
    {
        $params{args} = $request_body;
    }
    else # process multi-line data
    {
        # decode multi-part data
        if ($r_body[0] =~ /^-/)
        {
            my @request_parts;

            # trim whitespace on header
            $r_body[0] =~ s/\s//g;

            # grab segments
            my @parts = split(/$r_body[0]\-*\s*/, $request_body);
            @parts = grep { $_ } @parts;

            for my $part (@parts)
            {
                my @lines = split("\n", $part);

                $_ .= "\n" for @lines;

                # grab header
                my @header;
                for my $line (@lines)
                {
                    my $t_line = shift @lines;

                    $t_line =~ s/[\r\n]//g;
                    push @header, $t_line if $t_line;

                    last unless $t_line;
                }

                $lines[-1] =~ s/[\r\n]//g;
                my $data = join('', @lines);

                my $name;
                if ($header[0] =~ /name="(.*?)"/)
                {
                    $name = $1;
                }

                push @request_parts, {
                    name   => $name,
                    header => \@header,
                    data   => $data,
                };
            }
            $params{request_parts} = \@request_parts;
        }
    }

    return &init_dispatcher($r, %params);
}

sub init_dispatcher {
    my ($r, %params) = @_;
    
    %request_data = (
        headers       => { 'Content-Type' => 'text/html' },
        output        => q[],
        status        => 200,
        server_object => $r,
        request       => $params{request} || $r->request_body,
        args          => $params{args}    || $r->args,
        request_parts => $params{request_parts},
        begin_time    => [gettimeofday],
    );
    
    my $class      = $r->variable('call_class');
    my $bless      = $r->variable('bless');
    my $base_class = $r->variable('base_class');
    my $method     = $r->variable('call_method');

    $r->variable('call_class',  undef);
    $r->variable('bless',       undef);
    $r->variable('base_class',  undef);
    $r->variable('call_method', undef);

    my $self = { };
    $bless ? bless($self, $class) : bless($self);

    my $sub_call = "$class\::$method";
    if (UNIVERSAL::can($class, $method))
    { 
        no strict 'refs';

        # pre-run sub, if defined
        my $init_class = $base_class || $class;
        if (UNIVERSAL::can($init_class, 'init') and not $method eq 'error')
        {
            no strict 'refs';
            eval {
                local $SIG{__DIE__} = sub { &format_error(shift) };
                if ($bless)
                {
                    $self->init;
                }
                else
                {
                    my $prerun_sub = "$init_class\::init";
                    $prerun_sub->($self);
                }
            };

            if ($@ and $@ ne "nginx-exit\n")
            {
                if ($method eq 'error')
                {
                    # you've got an error in your error handler
                    warn "Error in error handler... ($class\::error)\n";

                    return $self->init_error($sub_call);
                }

                return &dispatch(
                    $self->server,
                    class  => $init_class,
                    method => 'error', 
                    error  => "$@",
                );
            }
        }

        my $error = $self->server->variable('error');

        if ($self->dispatching)
        {
            eval { 
                local $SIG{__DIE__} = sub { &format_error(shift) };

                if ($bless)
                {
                    $self->$method($error);
                }
                else
                {
                    $sub_call->($self, $error);
                }
            };

            if ($@ and $@ ne "nginx-exit\n")
            {
                if ($method eq 'error')
                {
                    # you've got an error in your error handler
                    warn "Error in error handler... ($class\::error)\n";

                    return $self->init_error($sub_call);
                }

                return &dispatch(
                    $self->server,
                    class  => $init_class,
                    method => 'error', 
                    error  => "$@",
                );
            }
            else
            {
                # process all data
                $self->process_auto_header
                    if $self->auto_header and $self->dispatching;

                # post-run sub, if defined
                my $cleanup_class = $base_class || $class;
                if (UNIVERSAL::can($cleanup_class, 'cleanup') and $method ne 'error'
                    and $self->dispatching)
                {
                    no strict 'refs';
                    eval { 
                        local $SIG{__DIE__} = sub { &format_error(shift) };
                        if ($bless)
                        {
                            $self->cleanup;
                        }
                        else
                        {
                            my $cleanup_sub = "$cleanup_class\::cleanup";
                            $cleanup_sub->($self);
                        }
                    };
                }
            }
        }

        undef $self;

        return OK;

    }
    else
    {
        return $self->init_error($sub_call);
    }
}

=head3 process_auto_header

Process the autoheader.

=cut

sub process_auto_header
{
    my $self = shift;

    $self->server->status($self->status);
            
    my $content_type = delete $request_data{headers}{'Content-Type'};

    $self->server->header_out($_, $request_data{headers}{$_})
        for keys %{$request_data{headers}};

    $self->server->send_http_header($content_type);

    $self->server->print($request_data{output});

    # ensure all data is transmitted
    $self->flush;
}

sub format_error
{
    my $error  = shift;
    my @stack  = &make_error_stack;
    $die_error = $error;

    return if $error eq "nginx-exit\n";

    warn $error;

    for my $e (@stack)
    {
        warn "$e->{sub} called at $e->{file} line $e->{line}\n";
    }
}

sub error_stack { @error_stack };
sub get_error   { $die_error   };

sub make_error_stack
{
    my @stack;
    my $i = 0;
    while (my @x = caller(++$i)) {
        push @stack, {
            pack => $x[0],
            file => $x[1],
            line => $x[2],
            sub  => $x[3],
        };
    }

    shift @stack;
    shift @stack;

    pop @stack;

    for (1..5)
    {
        pop @stack
            if $stack[-1] and $stack[-1]->{sub} =~ /^Nginx::Simple/;

        pop @stack
            if $stack[-1] and $stack[-1]->{file} =~ /\/Nginx\/Simple.pm$/;
    }

    @error_stack = @stack;

    return @stack;
}

sub init_error
{
    my ($self, $sub) = @_;
    
    warn(__PACKAGE__ . qq[: '$sub' does not exist...\n]) 
        unless $sub =~ /error$/;

    return HTTP_SERVER_ERROR;
}

=head3 $self->unescape

Unscape HTTP URI encoding.

=cut

sub unescape
{
    my ($self, $value) = @_;

    $value =~ s/\+/ /g;
    $value = $self->server->unescape($value);

    return $value;
}

=head3 $self->cookie

Cookie methods:

   $self->cookie->set(-name => 'foo', -value => 'bar');
   my %cookies = $self->cookie->read;

=cut

sub cookie { new Nginx::Simple::Cookie(shift) }

# override CORE::GLOBAL::exit & print
{
    no strict 'refs';
    *{"CORE::GLOBAL::exit"}  = sub { die "nginx-exit\n" };
    *{"CORE::GLOBAL::print"} = sub { };
}

=head3 $self->elapsed_time

Returns elapsed time since initial dispatch.

=cut

sub elapsed_time { tv_interval($request_data{begin_time}, [gettimeofday]) }

=head1 Author

Michael J. Flickinger, C<< <mjflick@open-site.org> >>

=head1 Copyright & License

You may distribute under the terms of either the GNU General Public
License or the Artistic License.

=cut

1;
